src/compiler/checker.ts

```diff
@@ -44538,21 +44538,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function checkTestingKnownTruthyCallableOrAwaitableOrEnumMemberType(condExpr: Expression, condType: Type, body?: Statement | Expression) {
         if (!strictNullChecks) return;
+
         bothHelper(condExpr, body);
 
         function bothHelper(condExpr: Expression, body: Expression | Statement | undefined) {
             condExpr = skipParentheses(condExpr);
-
             helper(condExpr, body);
-
             while (isBinaryExpression(condExpr) && (condExpr.operatorToken.kind === SyntaxKind.BarBarToken || condExpr.operatorToken.kind === SyntaxKind.QuestionQuestionToken)) {
                 condExpr = skipParentheses(condExpr.left);
                 helper(condExpr, body);
             }
         }
 
         function helper(condExpr: Expression, body: Expression | Statement | undefined) {
-            const location = isLogicalOrCoalescingBinaryExpression(condExpr) ? skipParentheses(condExpr.right) : condExpr;
+            let location = condExpr;
+            let inverted = false;
+            if (isPrefixUnaryExpression(location)) {
+                location = skipParentheses(location.operand);
+                inverted = true;

```

Comment:
Another thing to consider (and probably add a test) is that we can have an expression that flips between the negative and positive case, if we have nested negation expressions, e.g.: `if (a && !(b || !c)) ...`

Reply:
ah 🙈

Reply:
I don't think there's really any special handling needed here?